### ISSUE TITLE
GA support for reading from journald

### ISSUE THREAD

**Issue by cmacknz** [elastic/beats#37086]
- Relates https://github.com/elastic/elastic-agent/issues/3650

As of Debian 12 system logs are exclusively available via journald by default. Today we support reading journald logs via the Filebeat [journald input](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-journald.html), which is still in technical preview and has several major bugs filed against it. See https://github.com/elastic/beats/issues?q=is%3Aissue+is%3Aopen+journald notably:

- [ ] https://github.com/elastic/beats/issues/34077
- [ ] https://github.com/elastic/beats/issues/32782
- [ ] https://github.com/elastic/beats/issues/30398

We need to provide a GA way to read journald logs. There are two paths to this:

1. Fix the major issues in the journald input and GA it as is. All integrations that previously read syslog files by default will need a conditional to specify that journald should be used instead of one of the log files on Linux (see [example](https://github.com/elastic/integrations/blob/9dcd6eac196eed413eff898c1ecbc46fe1d745e1/packages/iptables/data_stream/log/agent/stream/journald.yml.hbs#L1). Possibly this conditional will need to be on the Linux distribution and not just Linux as a platform.
2. Fold the existing journald functionality into filestream, so that there is only one way to read log files and all existing uses of filestream to read system logs continue to work with no or minimal modification. In the ideal case we detect we are reading journald logs based on a .journal extension or well known file paths, but we may need a configuration flag for this. If we do end up with a configuration flag we could consider implementing journald support as a type of parser https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-filestream.html#_parsers

Option 1 is the easiest to do inside Beats, but will require every user and integration author to know they are reading journald logs and opt into it explicitly.

Option 2 is more work but makes journald support more seamless, and ensures features added to filestream benefit journald use cases where applicable.

**Comment by cmacknz** [comment]
@rdner I am interested to get your opinion on this given the amount of time you are spending trying to migrate and drive consistency between the log, filestream, and container input that already exist.

**Comment by leehinman** [comment]
For Option1 do we have to provide a conditional?  I think both inputs could be enabled at the same time, it would just have to be non-fatal for the source not to be present.  For example you can enable both journald, logfile & udp in iptables integration all at the same time.  (And UDP and journald are on by default)

**Comment by cmacknz** [comment]
If we don't have a conditional we risk duplicated logs. I think if we defaulted to always using both inputs we'd get a small amount of duplicated logs today on Debian 11, it looks like the kernal boot logs go to both journald and /var/log/

```sh
craig_mackenzie@cmackenzie-debian11-test:~$ journalctl
-- Journal begins at Tue 2023-11-14 20:15:17 UTC, ends at Tue 2023-11-14 20:19:35 UTC. --
Nov 14 20:15:17 debian kernel: Linux version 5.10.0-26-cloud-amd64 (debian-kernel@lists.debian.org) (gcc-10 (D>
Nov 14 20:15:17 debian kernel: Command line: BOOT_IMAGE=/boot/vmlinuz-5.10.0-26-cloud-amd64 root=UUID=62c0943b>
Nov 14 20:15:17 debian kernel: BIOS-provided physical RAM map:
Nov 14 20:15:17 debian kernel: BIOS-e820: [mem 0x0000000000000000-0x0000000000000fff] reserved
Nov 14 20:15:17 debian kernel: BIOS-e820: [mem 0x0000000000001000-0x0000000000054fff] usable
Nov 14 20:15:17 debian kernel: BIOS-e820: [mem 0x0000000000055000-0x000000000005ffff] reserved
Nov 14 20:15:17 debian kernel: BIOS-e820: [mem 0x0000000000060000-0x0000000000097fff] usable
Nov 14 20:15:17 debian kernel: BIOS-e820: [mem 0x0000000000098000-0x000000000009ffff] reserved
Nov 14 20:15:17 debian kernel: BIOS-e820: [mem 0x0000000000100000-0x00000000bf8ecfff] usable
Nov 14 20:15:17 debian kernel: BIOS-e820: [mem 0x00000000bf8ed000-0x00000000bf9ecfff] reserved
Nov 14 20:15:17 debian kernel: BIOS-e820: [mem 0x00000000bf9ed000-0x00000000bfaecfff] type 20
Nov 14 20:15:17 debian kernel: BIOS-e820: [mem 0x00000000bfaed000-0x00000000bfb6cfff] reserved
Nov 14 20:15:17 debian kernel: BIOS-e820: [mem 0x00000000bfb6d000-0x00000000bfb7efff] ACPI data
Nov 14 20:15:17 debian kernel: BIOS-e820: [mem 0x00000000bfb7f000-0x00000000bfbfefff] ACPI NVS
Nov 14 20:15:17 debian kernel: BIOS-e820: [mem 0x00000000bfbff000-0x00000000bffdffff] usable
Nov 14 20:15:17 debian kernel: BIOS-e820: [mem 0x00000000bffe0000-0x00000000bfffffff] reserved
Nov 14 20:15:17 debian kernel: BIOS-e820: [mem 0x0000000100000000-0x000000013fffffff] usable
Nov 14 20:15:17 debian kernel: printk: bootconsole [earlyser0] enabled
Nov 14 20:15:17 debian kernel: NX (Execute Disable) protection: active
Nov 14 20:15:17 debian kernel: efi: EFI v2.70 by EDK II
Nov 14 20:15:17 debian kernel: efi: TPMFinalLog=0xbfbf7000 ACPI=0xbfb7e000 ACPI 2.0=0xbfb7e014 SMBIOS=0xbf9ca0>
Nov 14 20:15:17 debian kernel: secureboot: Secure boot disabled
Nov 14 20:15:17 debian kernel: SMBIOS 2.4 present.
Nov 14 20:15:17 debian kernel: DMI: Google Google Compute Engine/Google Compute Engine, BIOS Google 10/09/2023
Nov 14 20:15:17 debian kernel: Hypervisor detected: KVM
Nov 14 20:15:17 debian kernel: kvm-clock: Using msrs 4b564d01 and 4b564d00
Nov 14 20:15:17 debian kernel: kvm-clock: cpu 0, msr 78801001, primary cpu clock
Nov 14 20:15:17 debian kernel: kvm-clock: using sched offset of 7655756989 cycles
Nov 14 20:15:17 debian kernel: clocksource: kvm-clock: mask: 0xffffffffffffffff max_cycles: 0x1cd42e4dffb, max>
Nov 14 20:15:17 debian kernel: tsc: Detected 2200.158 MHz processor
Nov 14 20:15:17 debian kernel: e820: update [mem 0x00000000-0x00000fff] usable ==> reserved
Nov 14 20:15:17 debian kernel: e820: remove [mem 0x000a0000-0x000fffff] usable
Nov 14 20:15:17 debian kernel: last_pfn = 0x140000 max_arch_pfn = 0x400000000
Nov 14 20:15:17 debian kernel: MTRR default type: write-back
Nov 14 20:15:17 debian kernel: MTRR fixed ranges enabled:
Nov 14 20:15:17 debian kernel:   00000-9FFFF write-back
Nov 14 20:15:17 debian kernel:   A0000-FFFFF uncachable
(reverse-i-search)`': ^C
craig_mackenzie@cmackenzie-debian11-test:~$ grep -rn 'kvm-clock: cpu 0, msr 78801001, primary cpu cloc' /var/log
grep: /var/log/journal/3465bc73197d954b92a16251605729f5/system.journal: binary file matches
grep: /var/log/private: Permission denied
grep: /var/log/btmp: Permission denied
/var/log/syslog:125:Nov 14 20:15:18 debian kernel: [    0.000000] kvm-clock: cpu 0, msr 78801001, primary cpu clock
/var/log/messages:27:Nov 14 20:15:18 debian kernel: [    0.000000] kvm-clock: cpu 0, msr 78801001, primary cpu clock
grep: /var/log/chrony: Permission denied
/var/log/kern.log:27:Nov 14 20:15:18 debian kernel: [    0.000000] kvm-clock: cpu 0, msr 78801001, primary cpu clock
```

Granted if someone set their logs path to `/var/log/*.log` they'd pick up these logs from both syslog.log and kern.log today anyway.

**Comment by cmacknz** [comment]
It also looks like the journald input is using go-systemd/sdjournal/ which is just wrapping the systemd journal C API:

https://github.com/coreos/go-systemd/blob/7d375ecc2b092916968b5601f74cca28a8de45dd/sdjournal/journal.go#L424-L434

```go
func NewJournal() (j *Journal, err error) {
	j = &Journal{}

	sd_journal_open, err := getFunction("sd_journal_open")
	if err != nil {
		return nil, err
	}

	r := C.my_sd_journal_open(sd_journal_open, &j.cjournal, C.SD_JOURNAL_LOCAL_ONLY)
```

This wouldn't fit with the idea of just using a filestream parser for journald, at best we could just hide the entire journald input inside filestream so there's a single log input, but we'd probably still need dedicated configuration specific to reading journald files.

**Comment by cmacknz** [comment]
The default journald configuration that reads everything is only two lines so I think at this point I'm convinced that keeping the journald input and improving it is the best path:

```yaml
# Read all journald logs
- type: journald
  id: everything
```

I don't think folding this into filestream will make filestream easier to use, or be easier to maintain.

**Comment by rdner** [comment]
To summarise what we discussed with @cmacknz on a call:

1. I think we should detect whether the OS has journald in the agent and add a new variable in the `host` object for the integration templates to use it like we use the condition here https://github.com/elastic/integrations/blob/f1b08ddd00724eaf3b8d9eb9ef2221f8fc7eefc4/packages/system/data_stream/system/agent/stream/winlog.yml.hbs#L2C1-L2C41

2. I think the users who use the agent are not interested in deep configuration, so the integrations should deal with logs in both files and journald seamlessly for the user.

3. Users who run standalone Filebeat are used to manual configurations and should be able to take care of configuring the right input for the right OS/distribution – journald or filestream.

4. journald should remain separate, it's not really compatible with the filestream architecture and consumes logs via special syscalls.

**Comment by andrewkroh** [comment]
A few things that come to mind related to journald:

- The input produces large events with lots of metadata. This could have an impact on storage usage. It also might make sense to drop some of the fields.
- The input is not optimized for producing ECS fields. IIRC it does populate ECS fields but it also duplicates the same data into non-ECS fields. It would be much better to optimize the events coming out of the input _before_ turning this on for users by default. All journald input users would benefit IMO.
- To make reading data most efficient for each data stream (system.syslog and system.auth) ideally we would use journalctl filtering (e.g. system.auth might use `_TRANSPORT=syslog`). So we need to figure out if that associated data is available in journald and what are the appropriate filters that can be used to select it. Implementing filter in the Beat using processors would be less that ideal for efficiency. (Viewing the data with `sudo journalctl -o export` is a great way to view the raw data).

**Comment by cmacknz** [comment]
Thanks, I think it would make sense to compare the events collected without the journald inputs to those collected with it. If the event content is significantly different it will cause problems for dashboards and queries.

**Comment by andrewkroh** [comment]
This is alluded to in some linked issues, but I wanted to explicitly mention that the journald library version in our container images is v245 (from Mar 6, 2020), and when deploying this image on Ubuntu 22.04 nodes, which uses v249, you can't collect logs from the host (no crashes, just no logs). My workaround has been to repack the filebeat binaries in with a more recent base image. We might want to consider bumping our base image as part of making this GA.

**Comment by belimawr** [comment]
I found another bug, probably another blocker: https://github.com/elastic/beats/issues/39352

It seems that if Filebeat falls too far behind with the Journal the input will crash shortly after starting.

**Comment by belimawr** [comment]
https://github.com/elastic/beats/issues/32782 and https://github.com/elastic/beats/issues/39352 happen intermittently on my test environments, so far I did not manage to isolate them but they both are coming from a call to `github.com/coreos/go-systemd/v22/sdjournal` https://github.com/elastic/beats/blob/ffcd1814666645a5d7a644911ecf6e2b7d8db3f5/filebeat/input/journald/pkg/journalread/reader.go#L185

https://github.com/elastic/beats/issues/39352 I only managed to reproduce with Journald `systemd 252 (252.16-1.amzn2023.0.2)`

**Comment by belimawr** [comment]
Now that we support reporting status per input when running under Elastic-Agent and given that Journald can cause Filebeat to crash (as detailed by https://github.com/elastic/beats/issues/34077), I added a task to ensure we have input status reporting for Journald before we GA so we can better report unsupported Systemd versions.

**Comment by belimawr** [comment]
Talking with @cmacknz the best course of action for those crashes is to use journalctl to read the journald logs instead of the current `github.com/coreos/go-systemd/v22/sdjournal`

I created an issue to track it: https://github.com/elastic/beats/issues/39820

**Comment by cmacknz** [comment]
Direct use of `journalctl` is the approach currently used by the OTel collector, related discussion in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/2332

**Comment by cmacknz** [comment]
From the comment in https://github.com/elastic/beats/issues/37086#issuecomment-1885856333 implying the documents the journald input produces are unlikely to be optimal today, we need to ask a task to compare the structure and metadata of the logs collected by journald to those collected today from syslog with the log input and ensure we only see expected differences.

**Comment by belimawr** [comment]
@cmacknz, @pierrehilbert I've been working on using `journalctl` to read journald logs and comparing the events generated by my POC implementation with our current implementation. With pretty little work, the generated events are the same, which is good news.

However, the filtering we have ([`include_matches`](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-journald.html#filebeat-input-journald-include-matches)) is pretty broken. The example in the documentation and the reference config file differ in syntax, none of them work as expected for our current implementation :cry: 

So far I could access that the problem is caused by two issues:
1. The way we represent and parse the config keys
2. The way we call `go-systemd` to add the filters

Bear in mind this `include_matches` is an advanced filtering, without using it we can already filter by unit, syslog identifier, facility, priority, match messages with grep-like patterns, etc.

Which brings me some product related questions:
1. What is the minimum feature set to GA journald?
2. Do we need to keep any of the current syntax for the advanced filtering?
3. Given Journald input is not GA and is not working as expected, how far can we go with breaking changes?

Just to give an example, the following YAML:
```yaml
include_matches:
  - and:
    - match:
      - FOO=bar
    - match:
      - BAR=bar
  - or:
    - match:
      - FOO_BAR=foo_bar
```

generates the following logic expression:
```
FOO_BAR=foo_bar OR FOO=foo AND BAR=bar AND
```

Yes, there is an `AND` without one of is operands, for some reason it does not break the input, but it does not produce the expected result.

I would expect the above YAML to produce the following logic expression:
```
(FOO=foo AND BAR=bar) OR (FOO_BAR=foo_bar)
```

Honestly, I don't think we should accept YAML above, it is confusing how the `AND` and `OR` relate to each other, which will lead users to make all sorts of mistakes.

**Comment by cmacknz** [comment]
The journald input is in technical preview, you can make whatever changes are needed breaking or not. That said, people are using it as is so don't make breaking changes that don't actually help.

Our end goal is not just to GA journald, it is to make Debian 12 based systems (and other distros that default to journald) work in our system integration. This means we want users to be able to get the same information out of journald that they would get out of syslog, without breaking any queries, dashboards, or alerts that already exist. So breaking changes need to be focused on the shape of the output data, and if it can be filtered in the same ways, and not on the input configuration syntax used to obtain it. 

If we can make it so the journald input is a drop in replacement for the log input in the system integration with no configuration changes that is even better, but I'm not sure this will be possible.

**Comment by belimawr** [comment]
Thanks @cmacknz, I'll take a look at the system integration on a Debian 12 host and what kind of filtering it uses for logs so I can draft the minimal requirements.

**Comment by belimawr** [comment]
TL;DR: The standard journald input, with no filters will collect all the data we need, to correctly add the dataset we can rely on the syslog facility code journald already adds to the events and are already published in the current version of the journald input.

So for https://github.com/elastic/beats/issues/39820 I'll focus on getting the core of the journald input working with `journalctl` with a good set of tests and leave advanced filtering to another task.

Long version:
I've been doing some testing with Debian 12 and research, here are the key findings:
**1. rsyslogd ships the data to the traditional log files**
The [Debian documentation](https://github.com/elastic/beats/issues/39820) states about the system message:
> Under systemd, the system logging utility rsyslogd(8) may be uninstalled. If it is installed, it changes its behavior to read the volatile binary log data (instead of pre-systemd default "/dev/log") and to create traditional permanent ASCII system log data. This can be customized by "/etc/default/rsyslog" and "/etc/rsyslog.conf" for both the log file and on-screen display. See rsyslogd(8) and rsyslog.conf(5).

The [Debian 12 Vagrant box](https://app.vagrantup.com/debian/boxes/bookworm64/versions/12.20240503.1) I've been using comes with `rsyslogd` installed and configured to forward the messages to the traditional log files, here is a snipped from `/etc/rsyslog.conf` where we can see the messages being directed to the files based on the log facility.

<details><summary>/etc/rsyslog.conf</summary>
<p>

```
#
# First some standard log files.  Log by facility.
#
auth,authpriv.*                 /var/log/auth.log
*.*;auth,authpriv.none          -/var/log/syslog
#cron.*                         /var/log/cron.log
daemon.*                        -/var/log/daemon.log
kern.*                          -/var/log/kern.log
lpr.*                           -/var/log/lpr.log
mail.*                          -/var/log/mail.log
user.*                          -/var/log/user.log

#
# Logging for the mail system.  Split it up so that
# it is easy to write scripts to parse these files.
#
mail.info                       -/var/log/mail.info
mail.warn                       -/var/log/mail.warn
mail.err                        /var/log/mail.err

#
# Some "catch-all" log files.
#
*.=debug;\
        auth,authpriv.none;\
        mail.none               -/var/log/debug
*.=info;*.=notice;*.=warn;\
        auth,authpriv.none;\
        cron,daemon.none;\
        mail.none               -/var/log/messages
```

</p>
</details> 

[Wikipedia lists](https://en.wikipedia.org/wiki/Syslog#Facility) the facilities and their names.

With that information it should be easy to "migrate" the system integration to use the journald input and test all ingest pipelines/dashboards.

**Comment by nerijus** [comment]
A lot of systems do not have rsyslog installed, so please don't make it a requirement for reading from journald.

**Comment by pierrehilbert** [comment]
I agree with @nerijus here, this is the main reason we need to support journald itself: avoiding to force our users to install rsyslog.

**Comment by belimawr** [comment]
> A lot of systems do not have (and do not want) rsyslog installed, so please don't make it a requirement for reading from journald.

We won't ;). I just mentioned rsyslog as an easy means of testing/comparing the traditional log files harvesting and the journald on the same system. We won't make rsyslog a dependency for the journald input.

**Comment by belimawr** [comment]
I'm not sure if that's critical to GA the journald input, but we should better support binary fields. So far I found them in two situations:
1. Messages with control characters: https://github.com/elastic/beats/blob/63fc18c71c2edc8fa56d66fe34d8c7c7a54dfdb0/filebeat/input/journald/pkg/journalctl/testdata/corner-cases.json#L64
2. Actual binary blobs: https://github.com/elastic/beats/blob/63fc18c71c2edc8fa56d66fe34d8c7c7a54dfdb0/filebeat/input/journald/pkg/journalctl/testdata/corner-cases.json#L14

Anyway I created an issue for that: https://github.com/elastic/beats/issues/40479

**Comment by andrewkroh** [comment]
> The input is not optimized for producing ECS fields. IIRC it does populate ECS fields but it also duplicates the same data into non-ECS fields. It would be much better to optimize the events coming out of the input before turning this on for users by default. All journald input users would benefit IMO.

I took a look at the [current state](https://github.com/elastic/beats/blob/63fc18c71c2edc8fa56d66fe34d8c7c7a54dfdb0/filebeat/input/journald/pkg/journalfield/default.go#L23) of the fields produced by the input. My suggestions for aligning closer to ECS are in this table. Once we settle on mappings we should add a table to the input's documentation.

| Current Name           | Proposed ECS Field       | Notes                                                                                                                                                                                                                                                |
|------------------------|--------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| message_id             | log.syslog.msgid         | https://www.elastic.co/guide/en/ecs/current/ecs-log.html#field-log-syslog-msgid                                                                                                                                                                      |
| syslog.priority        | log.syslog.priority      | https://www.elastic.co/guide/en/ecs/current/ecs-log.html#field-log-syslog-priority                                                                                                                                                                   |
| syslog.facility        | log.syslog.facility.code |                                                                                                                                                                                                                                                      |
| syslog.identifier      | log.syslog.appname       |                                                                                                                                                                                                                                                      |
| syslog.pid             | log.syslog.procid        |                                                                                                                                                                                                                                                      |
| systemd.cgroup         |                          | Leave as is?                                                                                                                                                                                                                                         |
| systemd.invocation_id  |                          | Leave as is?                                                                                                                                                                                                                                         |
| systemd.owner_uid      |                          | Leave as is?                                                                                                                                                                                                                                         |
| systemd.session        |                          | Leave as is?                                                                                                                                                                                                                                         |
| systemd.slice          |                          | Leave as is?                                                                                                                                                                                                                                         |
| systemd.unit           |                          | Leave as is?                                                                                                                                                                                                                                         |
| systemd.user_slice     |                          | Leave as is?                                                                                                                                                                                                                                         |
| systemd.user_unit      |                          | Leave as is?                                                                                                                                                                                                                                         |
| systemd.transport      |                          | Leave as is?                                                                                                                                                                                                                                         |
| container.id_truncated |                          | Already have container.id, not need to also have id_truncated. Drop this field.                                                                                                                                                                      |
| container.log.tag      |                          | Don't need it because it is duplicated into the SYLOG_IDENTIFER (https://docs.docker.com/engine/logging/drivers/journald/). Drop this field.                                                                                                         |
| container.partial      |                          | Means that the message was truncated. The value is a boolean "true". The input could append("tags", "partial_message") if CONTAINER_PARTIAL_MESSAGE == "true". We should not be adding fields into the container namespace that are not part of ECS. |

**Comment by belimawr** [comment]
I found an issue with our current (`main`, `8.x`) journald input implementation: https://github.com/elastic/beats/issues/41083

I've already added it to the must have list for GA. The previous implementation did not have this problem. I already have an idea of how to fix it and how to test it.

**Comment by belimawr** [comment]
> I think this was handled in other places by defining a forwarded tag that excludes the event from the add_host_metadata default processor. See [here](https://github.com/elastic/beats/blob/bfde79f3ddd1164dcb6e9d93e904dab79f47c02f/filebeat/filebeat.yml#L171-L173) and [here](https://github.com/elastic/beats/blob/bfde79f3ddd1164dcb6e9d93e904dab79f47c02f/x-pack/filebeat/cmd/root.go#L56-L59) for agent.

I saw that, that's also what I recommended in the disucss thread, however I'm not sure what is the best approach to "solve" this issue: Do we add the `forwarded` tag by default in the integration? Do we just add a toggle for it and document the behaviour?

My gut feeling tells me we should use different fields, at least in this context of one Filebeat ingesting logs from multiple hosts.

I'll centralise this discussion on a single issue because it's affecting multiple projects in different ways.

**Comment by jlind23** [comment]
Removed https://github.com/elastic/beats/issues/42208 from the nice to have and moved it to the must have section of this issue.

**Comment by belimawr** [comment]
I found bug in our Filebeat docker image: we're not shipping the `journalctl` binary that is required to run the Journald input.

There is a GH issue (https://github.com/elastic/beats/issues/41278) opened a while ago, technically it's not a bug with the input implementation, however it's a bad user experience for people running Filebeat in Docker/Kubernetes to collect Journald logs.

**Comment by belimawr** [comment]
It also makes me think how much of a supported/official use case is to run Filebeat in a docker container collecting journald logs 🤔 .

**Comment by cmacknz** [comment]
People would want this for reading Kubernetes node logs I think, for one use case I can think of off the top of my head. https://github.com/elastic/beats/issues/41278#issuecomment-2698344395 is a reasonable argument to just have journalctl in the base elastic-agent container.

There is going to be a -slim variant of agent soon (https://github.com/elastic/elastic-agent/pull/7173), we could put journalctl in elastic-agent and elastic-agent-complete but omit from -slim, but it depends on much it increases the container size.

**Comment by belimawr** [comment]
Reading Kubernetes node logs makes sense. We definitely should add test case for it (reading Journald logs from a container) on both repos: Beats and Elastic-Agent.

The Elastic-Agent might already have it indirectly because we test the `Custom Journald logs` integration.

We also need to better document the differences between the Elastic-Agent images, [our docs](https://www.elastic.co/guide/en/fleet/current/elastic-agent-container.html) are not clear about their differences.

**Comment by belimawr** [comment]
@nimarezainia, @pierrehilbert another thing that came up is about compatibility between the `journalctl` version we ship in our Docker images and the host user's are running.

A newer version of `journalctl` will read older journal formats, but the opposite is not always true.

Essentially this adds another dimension for the compatibility matrix of the Journald input from our Docker images: the `journalctl` version we're shipping.

Ex: The Elastic-Agent `v8.17.3` complete Docker image is based on Ubuntu 20.04, so (according to user reports) it is incompatible with the journal version used by GKE instances running Ubuntu 22.04.

**Comment by cmacknz** [comment]
In 9.0 we don't even ship ubuntu containers, what happens to journald compatibility then? The base images will be wolfi and RedHat UBI.

Agent is upgraded to Ubuntu 24.04 in 8.x (except that we undid it briefly) and your comment implies we always want to be on the latest stable Ubuntu release that is in our support matrix.

**Comment by belimawr** [comment]
> In 9.0 we don't even ship ubuntu containers, what happens to journald compatibility then? The base images will be wolfi and RedHat UBI.
> 
> Agent is upgraded to Ubuntu 24.04 in 8.x (except that we undid it briefly) and your comment implies we always want to be on the latest stable Ubuntu release that is in our support matrix.

Effectively it is not about the base image version, it is about the `journalctl` version we ship, that is what defines the compatibility with different distributions.

The latest `9.0.0-SNAPSHOT` ships with `systemd 245 (245.4-4ubuntu3.23)`
```
elastic-agent@e57706d8a7e2:~$ journalctl --version                                                                                                                                                                                                                           
systemd 245 (245.4-4ubuntu3.23)
```

The [release notes for Ubuntu 24.04](https://discourse.ubuntu.com/t/ubuntu-24-04-lts-noble-numbat-release-notes/39890#p-99950-systemd-v2554) states it is using `systemd v255.4`. So there can be some incompatibility.

I'd say we always want to ship a `journalctl` that is compatible (equal or greater version) the the with all distros in our support matrix. Without looking what each distro ships, Ubuntu 24.04 seems to be the one with the newest version of it.

**Comment by cmacknz** [comment]
> I'd say we always want to ship a journalctl that is compatible (equal or greater version) the the with all distros in our support matrix. Without looking what each distro ships, Ubuntu 24.04 seems to be the one with the newest version of it.

This makes sense to me, can you figure out the best way for this to happen consistently across all our docker base images?

**Comment by belimawr** [comment]
> This makes sense to me, can you figure out the best way for this to happen consistently across all our docker base images?

Yes, I can look into that.

The only question I have is the priority of it among my other sprint tasks.

**Comment by lpeter91** [comment]
I'll drop here some misc experiences/opinions about reading the Journal with the Agent. I know this is not a discussion forum, but I had limited luck on the Discourse in the past. Feel free to remove this if it's really out of place.

- Systemd has been the default init on most major distros for at least 10 years now. This means that Journald is now the de-facto logging interface on most Linux systems. So I think the current condition on whether to read system logs from the journal or the legacy log files is way to conservative. I think the journal should be preferred whenever it's available.
- Sadly checking if `/var/log/journal` exists is insufficient, since the current implementation is incompatible with some older versions of `journalctl` that does not yet have the `--facility` option. This is especially problematic on hosts where the legacy log files are long gone but still `journalctl` is not recent enough. Some Redhat 8 releases satisfy these criteria, so this might be an issue worth looking into. (The exact error is: `Journalctl wrote to stderr: journalctl: unrecognized option '--facility'`)
- A (very) niche use-case for the containerized Agent could be to support such hosts; but indeed the major use-case is reading logs from Kubernetes nodes.
- In general the new implementation is working well, and it's nice to see that it's now also part of the System integration.

**Comment by belimawr** [comment]
> What's the latest on this ?
> 
> I've tried to use the docker filebeat versions 8.18.0 and 9.0.0 which are now shown as GA and can't get either of them to work with journald input. Seems that journalctl is still not included in the image.

We're working on shipping the `journalctl` with our base docker images, you can follow this issue: https://github.com/elastic/beats/issues/44040

**Comment by belimawr** [comment]
> Next question is whether version 9.0.0 handles corrupted journal files without stalling

Could you provide an example of how to reproduce/test it?

Are you sure the files are corrupted or is it just a case of the journal files being generated by a newer version of journald than the `journalctl` in the docker image?

Btw, I'll be off for about 1.5 week, so it will take a while for me to get back to you.

**Comment by cmacknz** [comment]
https://github.com/elastic/beats/pull/40558 mitigated this by restarting journald automatically when this happens so it recovers. We are now just running journalctl directly to read the journald logs so your system will have whatever problems journald itself has, but we are no longer going to give up immediately on error.
